### ISSUE TITLE
mtl/ofi: Error returned by request event callback: 4.

### ISSUE THREAD

**Issue by darrylabbate** [open-mpi/ompi#11774]
Thank you for taking the time to submit an issue!

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

5.1.0a1 (`main`)

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

Source (Git clone)

### If you are building/installing from a git clone, please copy-n-paste the output from `git submodule status`.

```
 22fe51cb7a961b6060fc5c48e659237cbe162566 3rd-party/openpmix (v1.1.3-3872-g22fe51cb)
 0e8a082238d8c3ef544cc1e0bb151466cc883667 3rd-party/prrte (psrvr-v2.0.0rc1-4635-g0e8a082238)
 c1cfc910d92af43f8c27807a9a84c9c13f4fbc65 config/oac (heads/main)
```

### Please describe the system on which you are running

* Operating system/version: Ubuntu 20.04 (also seen on CentOS 7)
* Computer hardware: Amazon EC2 - c5n.18xlarge (Intel x86)
* Network type: EFA

-----------------------------

## Details of the problem

When running `osu_latency_mt` (multithreaded latency benchmark) on a single node, the benchmark sometimes fails in the middle of the run with the following message:

```
[c5n-no-mt-st-n72-1:367349] ../../../../../../ext/ompi/ompi/mca/mtl/ofi/mtl_ofi.h:158: Error returned by request event callback: 4.
*** The Open MPI OFI MTL is aborting the MPI job (via exit(3)).
```

If I modify the code to `abort()` instead of `exit(1)`, the following stack trace is produced:

```
[c5n-no-mt-st-n72-1:367349] *** Process received signal ***
[c5n-no-mt-st-n72-1:367349] Signal: Aborted (6)
[c5n-no-mt-st-n72-1:367349] Signal code:  (-6)
[c5n-no-mt-st-n72-1:367349] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x14420)[0x7ff16e07e420]
[c5n-no-mt-st-n72-1:367349] [ 1] /lib/x86_64-linux-gnu/libc.so.6(gsignal+0xcb)[0x7ff16debb00b]
[c5n-no-mt-st-n72-1:367349] [ 2] /lib/x86_64-linux-gnu/libc.so.6(abort+0x12b)[0x7ff16de9a859]
[c5n-no-mt-st-n72-1:367349] [ 3] /home/ubuntu/.local/ompi/lib/libmpi.so.0(+0x1ec319)[0x7ff16e279319]
[c5n-no-mt-st-n72-1:367349] [ 4] /home/ubuntu/.local/ompi/lib/libmpi.so.0(+0x28e25b)[0x7ff16e31b25b]
[c5n-no-mt-st-n72-1:367349] [ 5] /home/ubuntu/.local/ompi/lib/libmpi.so.0(MPI_Send+0x183)[0x7ff16e166a13]
[c5n-no-mt-st-n72-1:367349] [ 6] /home/ubuntu/.local/ompi/libexec/osu-micro-benchmarks/mpi/pt2pt/osu_latency_mt(+0x3112)[0x55fdb6996112]
[c5n-no-mt-st-n72-1:367349] [ 7] /lib/x86_64-linux-gnu/libpthread.so.0(+0x8609)[0x7ff16e072609]
[c5n-no-mt-st-n72-1:367349] [ 8] /lib/x86_64-linux-gnu/libc.so.6(clone+0x43)[0x7ff16df97133]
[c5n-no-mt-st-n72-1:367349] *** End of error message ***
--------------------------------------------------------------------------
prterun noticed that process rank 1 with PID 367349 on node c5n-no-mt-st-n72-1 exited on signal 6 (Aborted).
--------------------------------------------------------------------------
```

However if I compile a debug build (`--enable-debug`), I cannot reproduce the failure at all (after thousands of iterations).

I suspect some downstream function call is returning an uninitialized variable. Sometimes the error message prints out a code other than `4`:

```
mtl_ofi.h:158: Error returned by request event callback: -2080370688.
```

GCC spits out a good number of warning for "maybe uninitialized" variables (mostly for PRRTE). Ex:

```
../../../../../ext/ompi/3rd-party/prrte/src/prted/pmix/pmix_server.c: In function ‘pmix_server_init’:
../../../../../ext/ompi/3rd-party/prrte/src/prted/pmix/pmix_server.c:594:16: warning: ‘rc’ may be used uninitialized in this function [-Wmaybe-uninitialized]
  594 |         return rc;
      |                ^~
../../../../../ext/ompi/3rd-party/prrte/src/prted/pmix/pmix_server.c: In function ‘pmix_server_dmdx_recv’:
../../../../../ext/ompi/3rd-party/prrte/src/prted/pmix/pmix_server.c:1371:9: warning: ‘rc’ may be used uninitialized in this function [-Wmaybe-uninitialized]
 1371 |         send_error(rc, &pproc, sender, index);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I'm also generally curious about PRRTE/PMIX since it spits out this error message upon running the benchmark:

```
PMIX ERROR: ERROR in file ../../../../../ext/ompi/3rd-party/openpmix/src/client/pmix_client_topology.c at line 352
```

Full shell command:

```shell
mpirun \
  -n 2 \
  --report-bindings \
  --map-by ppr:2:node \
  --rank-by slot \
  --oversubscribe \
  /path/to/osu_latency_mt
```

I was also able to reproduce this failure with just `--map-by package --bind-to package` (without `--oversubscribe`).

Possibly related:

- #11541
- #11726

Any help with debugging is appreciated

**Comment by wenduwan** [comment]
We temporarily mitigated this. Need to investigate later.

**Comment by darrylabbate** [comment]
Should be resolved by #12175

**Comment by darrylabbate** [comment]
#12175 merged
